Cython/Compiler/ModuleNode.py

```diff
@@ -2336,6 +2432,149 @@ def generate_import_star(self, env, code):
         code.putln(UtilityCode.load_as_string("ImportStar", "ImportExport.c")[1])
         code.exit_cfunc_scope()  # done with labels
 
+    def generate_module_state_start(self, env, code):
+        code.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
+        code.putln('typedef struct {')
+        code.putln('PyObject *%s;' % Naming.builtins_cname)
+        code.putln('PyObject *%s;' % Naming.cython_runtime_cname)
+        code.putln('PyObject *%s;' % Naming.empty_tuple)
+        code.putln('PyObject *%s;' % Naming.empty_bytes)
+        code.putln('PyObject *%s;' % Naming.empty_unicode)
+        if Options.pre_import is not None:
+            code.putln('PyObject *%s;' % Naming.preimport_cname)
+        code.putln('int %s;' % Naming.lineno_cname)
+        code.putln('int %s;' % Naming.clineno_cname)
+        code.putln('const char *%s;' % Naming.filename_cname)
+
+    def generate_module_state_end(self, env, modules, globalstate):
+        module_state = globalstate['module_state']
+        module_state_defines = globalstate['module_state_defines']
+        module_state_clear = globalstate['module_state_clear']
+        module_state_traverse = globalstate['module_state_traverse']
+        for module in modules:
+            definition = module is env
+            for entry in env.c_class_entries:
+                if definition or entry.defined_in_pxd:
+                    module_state.putln("PyObject *%s;" % entry.type.typeobj_cname)
+                    module_state_defines.putln("#define %s %s->%s" % (
+                        entry.type.typeobj_cname,
+                        Naming.modulestateglobal_cname,
+                        entry.type.typeobj_cname))
+                    module_state_clear.putln("  Py_CLEAR(%s(m)->%s);" % (
+                        Naming.modulestate_cname,
+                        entry.type.typeobj_cname))
+                    module_state_traverse.putln("  Py_VISIT(%s(m)->%s);" % (
+                        Naming.modulestate_cname,
+                        entry.type.typeobj_cname))
+        module_state.putln('} %s;' % Naming.modulestate_cname)
+        module_state.putln('')
+        module_state.putln('#ifdef __cplusplus')
+        module_state.putln('namespace {')
+        module_state.putln('extern struct PyModuleDef %s;' % Naming.pymoduledef_cname)
+        module_state.putln('} /* anonymous namespace */')
+        module_state.putln('#else')
+        module_state.putln('static struct PyModuleDef %s;' % Naming.pymoduledef_cname)
+        module_state.putln('#endif')
+        module_state.putln('')
+        module_state.putln('#define %s(o) ((%s *)PyModule_GetState(o))' % (
```

Comment:
Is it really guaranteed that this cannot fail? (Same for `PyState_FindModule()` below.)

Reply:
This can indeed fail and return `NULL`. However, this should work just like a global being `NULL`, therefore I explicitly didn't add error handling here. The expectation is that wherever this is used, it will have error handling

Reply:
Then let me add that none of the usages that I see in your implementation (all the calls to `Py_CLEAR()`, `Py_VISIT()`, etc.) has such a NULL check nor error handling, and will thus lead to a crash if the call ever fails. I think we should avoid calling this too lightly and often, and also try to make it more visible that this macro is much less trivial than it looks.

Reply:
> and will thus lead to a crash if the call ever fails

Both `Py_CLEAR` and `Py_VISIT` are able to handle NULL values, it's a no-op in that case. Did I miss any other cases? If so, I could surely add error handling to those!

> I think we should avoid calling this too lightly and often, and also try to make it more visible that this macro is much less trivial than it looks.

Any suggestions? What about getting rid of the `Naming.modulestate_cname` macro and add the `PyModule_GetState` usage directly? Thoughts?

Reply:
I added my suggestion of using a `__Pyx_PyModule_GetState` to check the module state before returning. Although I wonder if we can handle this more gracefully than just a Py_FatalError, let me know what you think.